field_type.go

```diff
@@ -84,7 +83,7 @@ func (s *scalarT) Element() FieldTypeElem { return nil }
 func (s *scalarT) Key() FieldTypeElem     { return nil }
 
 func (s *scalarT) IsOptional() bool {
-	return !s.fld.Syntax().SupportsRequiredPrefix() || s.ProtoLabel() == Optional
+	return s.fld.HasOptionalKeyword()
```

Comment:
This would make all proto3 fields without presence suddenly become `IsOptional() == false` which is very backwards incompatible. 

Proto3 optional label is not the same as proto2. The proto2 label means "this field is not required" whereas the proto3 label means "this field is still not required, but it has explicit presence semantics". 

This `IsOptional` in particular is referring to the field not being required, which is not affected by the explicit presence semantics. So IIUC this method does not need to change.